### PR TITLE
[Scanner] Register current execution scanner

### PR DESCRIPTION
**Pull request by mkolodezny**
**How did you test it?**
unit tests
run both enabled scanners locally:

```
cadence % ./cadence --do  cadence-system  wf list  --w cadence-sys-current-executions-scanner listall
             WORKFLOW TYPE            |              WORKFLOW ID               |                RUN ID                |                     TASK LIST                     | START TIME | EXECUTION TIME | END TIME
  ...rent-executions-scanner-workflow | cadence-sys-current-executions-scanner | 26b9991f-c44b-4e64-bd22-455caa1710df | cadence-sys-current-executions-scanner-tasklist-0 | 12:05:10   | 12:06:00       | 12:21:58
  ...rent-executions-scanner-workflow | cadence-sys-current-executions-scanner | 4e77431d-bac9-4c1f-8786-7062843e7521 | cadence-sys-current-executions-scanner-tasklist-0 | 12:04:11   | 12:05:00       | 12:05:10
  ...rent-executions-scanner-workflow | cadence-sys-current-executions-scanner | bfc715cf-f523-4813-9f34-aaeb7e74990b | cadence-sys-current-executions-scanner-tasklist-0 | 12:03:10   | 12:04:00       | 12:04:11
........................

cadence % ./cadence --do  cadence-system  wf list  --w cadence-sys-executions-scanner listall
             WORKFLOW TYPE            |          WORKFLOW ID           |                RUN ID                |                 TASK LIST                 | START TIME | EXECUTION TIME | END TIME
  ...-sys-executions-scanner-workflow | cadence-sys-executions-scanner | ee98f9eb-eb74-4536-a494-d5ae02b52b58 | cadence-sys-executions-scanner-tasklist-0 | 12:15:03   | 12:16:00       | 12:22:08
  ...-sys-executions-scanner-workflow | cadence-sys-executions-scanner | 3ce3e68c-4320-4998-b30b-040e998f73ac | cadence-sys-executions-scanner-tasklist-0 | 11:55:06   | 11:56:00       | 12:15:03
 ...-sys-executions-scanner-workflow | cadence-sys-executions-scanner | 0a8a20a7-2361-4b6a-ad48-ca5bff718cfd | cadence-sys-executions-scanner-tasklist-0 | 11:39:35   | 11:40:00       | 11:55:06
.......................
```

This is NOT the last change for the scanner.

### CODE REVIEW

File: service/worker/scanner/scanner.go [hunk 1]
```diff
@@ -117,18 +119,16 @@ func New(
...
-			Resource:                     s.context.Resource,
-			Scope:                        s.context.Resource.GetMetricsClient().Scope(metrics.ExecutionsScannerScope),
-			ScannerWorkflowDynamicConfig: s.context.cfg.ExecutionScannerConfig,
-		})
-	}
-	backgroundActivityContext = context.WithValue(backgroundActivityContext, executions.FixerContextKey, executions.FixerContext{
-		Resource: s.context.Resource,
-		Scope:    s.context.Resource.GetMetricsClient().Scope(metrics.ExecutionsFixerScope),
-	})
+	backgroundActivityContext, taskListNames :=
```
Comment (vancexu) [inline service/worker/scanner/scanner.go:122]: why taskListNames is needed to be returned from startExecutionWorkflowWithRetry?

File: service/worker/scanner/executions/activities.go [hunk 2]
```diff
@@ -361,7 +364,7 @@ func FixShardActivity(
 	for i := heartbeatDetails.LastShardIndexHandled + 1; i < len(params.CorruptedKeysEntries); i++ {
 		currentShardID := params.CorruptedKeysEntries[i].ShardID
 		currentKeys := params.CorruptedKeysEntries[i].CorruptedKeys
-		shardReport, err := fixShard(activityCtx, params, currentShardID, currentKeys, heartbeatDetails)
+		shardReport, err := fixShard(activityCtx, params, currentShardID, currentKeys, heartbeatDetails, params.ScanType)
```
Comment (vancexu) [inline service/worker/scanner/executions/activities.go:367]: looks like params is already passed in

File: common/reconciliation/invariants/invariantManager.go [hunk 3]
```diff
@@ -127,14 +128,15 @@ func (i *invariantManager) nextCheckResultType(
...
 	var ivs []common.Invariant
 	for _, collection := range collections {
 		switch collection {
 		case common.InvariantCollectionHistory:
-			ivs = append(ivs, getHistoryCollection(pr)...)
+			ivs = append(ivs, getHistoryCollection(pr, scanType)...)
 		case common.InvariantCollectionMutableState:
-			ivs = append(ivs, getMutableStateCollection(pr)...)
+			ivs = append(ivs, getMutableStateCollection(pr, scanType)...)
 		}
```
Comment (yycptt) [inline common/reconciliation/invariants/invariantManager.go:140]: How about creating a new invariant collection specifically for current mutable state records? I think it'll be easier for adding new checks for current records and `scanType` parameter won't be necessary.

File: service/worker/scanner/executions/activities.go (hunk 2 above)
Comment (mkolodezny) [reply service/worker/scanner/executions/activities.go:367]: good catch! The same thing with the scanShard too. Removed.

File: service/worker/service.go [hunk 4]
```diff
@@ -134,15 +134,26 @@ func NewConfig(params *service.BootstrapParams) *Config {
...
+			},
+			CurrentExecutionScannerConfig: &executions.ScannerWorkflowDynamicConfig{
+				Enabled:                 dc.GetBoolProperty(dynamicconfig.CurrentExecutionsScannerEnabled, false),
+				Concurrency:             dc.GetIntProperty(dynamicconfig.CurrentExecutionsScannerConcurrency, 25),
+				ExecutionsPageSize:      dc.GetIntProperty(dynamicconfig.CurrentExecutionsScannerPersistencePageSize, 1000),
+				BlobstoreFlushThreshold: dc.GetIntProperty(dynamicconfig.CurrentExecutionsScannerBlobstoreFlushThreshold, 100),
+				ActivityBatchSize:       dc.GetIntProperty(dynamicconfig.CurrentExecutionsScannerActivityBatchSize, 25),
+				DynamicConfigInvariantCollections: executions.DynamicConfigInvariantCollections{
+					InvariantCollectionMutableState: dc.GetBoolProperty(dynamicconfig.CurrentExecutionsScannerInvariantCollectionMutableState, true),
+					InvariantCollectionHistory:      dc.GetBoolProperty(dynamicconfig.ConcreteExecutionsScannerInvariantCollectionHistory, false),
```
Comment (yycptt) [inline service/worker/service.go:156]: dynamicconfig.**Current**ExecutionsScannerInvariantCollectionHistory?

File: service/worker/scanner/scanner.go (hunk 1 above)
Comment (mkolodezny) [reply service/worker/scanner/scanner.go:122]: We dont know those tasklists before call.

Comment (yycptt) [reply service/worker/scanner/scanner.go:122]: I think we need the names to create workers to poll from those tasklists.

**Comment by mkolodezny** [comment]
> Seems like this is the last diff for current record scanner? May I ask if we still plan to do this: [#3402 (comment)](https://github.com/uber/cadence/pull/3402#discussion_r459145950)

No, thats not the last diff for the scanner. The refactoring seems to be big enough to go separately. And it will be done with the load testing in parallel.

File: common/reconciliation/invariants/invariantManager.go (hunk 3 above)
Comment (mkolodezny) [reply common/reconciliation/invariants/invariantManager.go:140]: This class will be gone with the refactoring phase, as discussed, and so the parameter here as well.

**Pull request merged**